### ISSUE TITLE
Atomic Document Update Operations

### ISSUE THREAD

**Issue by merabtenei** [appwrite/appwrite#3905]
### 🔖 Feature description

There is a clear need of having some sort of atomic transactions at least for updating a field value based on its current value, like Incrementing or decrementing a field's value on the database.

### 🎤 Pitch

An example for this use case is Counters. In a blog post case we may want to count the number of likes or comments on every user operation (like or comment) by incrementing a counter (ex: likes_count, comments_count).
We can do that using server functions at the moment but there is no way to prevent race condition. (https://en.wikipedia.org/wiki/Race_condition)

Firestore have some solutions for this using either transactions or using an update based on the field's value.
```js
update({
'likes_count': FieldValue.increment(1),
})
```

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/appwrite/appwrite/blob/HEAD/CODE_OF_CONDUCT.md)

**Comment by Meldiron** [comment]
This became really requested feature recently. @eldadfux would it be possible to incorporate this into our near-future release plans?

---

@merabtenei This is what I did in my project, it may help solve your case in the meantime:
I have discussions which nest discussions recursively. Each discussion is either positive or negative. For each discussion, I need to know positive&negative ratio of all sub-discussions.

I created a function that updates counters on all parent discussions: https://github.com/Meldiron/almost-kialo/blob/master/backend/functions/onDiscussionCreate/src/index.js

This function now has the same problem as you - I am getting and updating a document, and there is a possibility of race condition: https://github.com/Meldiron/almost-kialo/blob/master/backend/functions/onDiscussionCreate/src/index.js#L65-L75

This only happens rarely with huge traffic, so I was not too worried. Also, it's must tiny metadata, not a huge feature.

Either way, I write a function that is triggered by CRON every 60 minutes. This function goes over all discussions and calculates their upvotes and downvotes based on how many discussions there are. This is a pretty slow function, tho it gets everything back in sync: https://github.com/Meldiron/almost-kialo/blob/master/backend/functions/aggregateVotes/src/index.js

These two combined create a counting solution that is always only a few numbers off, if not exact. I understand if it feels like extra work, and that is why I fully support this feature idea.

**Comment by merabtenei** [comment]
@Meldiron I totally agree, this may not happen with small traffic, but we can not rely on this, if the project gets big it would be really hard to change things from there.
Same for you solution, if we get to millions of likes or comments that function may take forever to finish.

**Comment by eldadfux** [comment]
I think adding increment and decrement functionality to the database makes lots of sense. I wonder if we should support this by introducing custom syntax when creating a new document or by adding a new column type.

New syntax could look something like this: 

```json
{
    "metric": "visits",
    "value": "increment(2)"
}
```

Problem: this completely messes type safety as `value`  should be an integer. We could overcome this internally with custom validators, and some SDKs helpers on the client side like we have for queries, permissions and roles. Might help us avoid having too many custom attributes.

The other option is to set a new special columns called `count` that can accepts a positive or negative value. We could potentially add support for both integer a float counters.

Would love more feedback.

**Comment by merabtenei** [comment]
@eldadfux To start couldn't we just add a way to use atomic database transaction on the server side functions ? AppWrite uses MariaDB if I'm not mistaken and MariaDB supports transaction natively. I think that wouldn't be complicated to add isn't it ?

**Comment by Meldiron** [comment]
@eldadfux Special column sounds limiting. We would need one for integer, one for float, and possibly one for string. Not to mention there would be no way to set the number to a specific initial value, for instance, when creating a document (`moneyBalance` in games would be a great example).

I like the approach with SDK helpers and `increment()` method. It will be harder to implement, but I find it useful anyway to start such a validator since it will be needed sooner or later for aggregation - for example, `sum()` or `count()`.

What do you think about supporting this for string attributes too? For example `javascript` and `increment(" python")` would result in value `javascript python` (example with tags system that is searchable). In SQL I believe this is supported, it's just matter of how we expose it. `increment` doesn't sound like a good name.

**Comment by eldadfux** [comment]
While transactions are in the plan, I find it hard to see how they could help solve this issue. The race conditions could still happen, and you can't really wrap all function's executions under a single transaction.

**Comment by eldadfux** [comment]
@Meldiron we could offer more functions, but I agree, for strings `increment` sounds like a bad name. I have one more problem with `increment()`, how will we implement range validation? If we want to use the DB atomic operation we won't be able to validate the the incremented result is still in valid range for the value.

We could potentially try to get the old value and do the math on the validators side, but I'm not sure if the result is going to be 100% accurate. This will have to be carefully tested.

**Comment by merabtenei** [comment]
@eldadfux is there any update on this ? Is there anything planned regarding this feature ? Could we see increment and decrement features soon ?

**Comment by stnguyen90** [comment]
Related PR: https://github.com/utopia-php/database/pull/217

**Comment by LDY681** [comment]
> Related PR: [utopia-php/database#217](https://github.com/utopia-php/database/pull/217)

Hi, excellent work! It looks to me that this is only for updating current document attributes.
This would struggles in situation such as creating a new document with incremented ID. 
There are many user cases such as creating a new article with newest article ID + 1.
Can we also have a API method ID.increment() similar to ID.unique()? 
Thanks!

**Comment by PratikDev** [comment]
is there any update about it? or anything we can use temporarily? I kinda need this feature urgently

**Comment by bazrafkan** [comment]
Any update? this feature is really essential.

**Comment by AH2023** [comment]
I got the value of the attribute beforehand using listDocuments and just added or subtracted 1. For example: 

        await database.updateDocument(databaseId, collectionId, documentId, {
          coins: bankDoc.coins -1
        })
        
bankDoc would be the required document that needs to be updated and bankDoc.coins would be the current value of the attribute that needs to be updated. 

This method likely won't work for a value that changes quite fast such as the number of likes, especially because many people are changing the value at roughly the same time so a locking mechanism would be required to prevent an error. However, I think appwrite does have a locking mechanism for situations like these but I'm not sure

**Comment by stnguyen90** [comment]
@PratikDev, please make sure you 👍🏼 this issue.

**Comment by sheenrep** [comment]
@PratikDev any update on that issue. We really need that to migrate from firebase completely.

**Comment by merabtenei** [comment]
We need at least incrementing and decrementing values safely which can be achieved with a server side function.
After that many use cases would need a bloc of instructions that can be executed atomically and can be reversed on failure.

**Comment by bazrafkan** [comment]
We also just need incrementing and decrementing atomic operations, @stnguyen90 Do you have any plan for this issue?

**Comment by nev-21** [comment]
> While transactions are in the plan, I find it hard to see how they could help solve this issue. The race conditions could still happen, and you can't really wrap all function's executions under a single transaction.

Hi @eldadfux @fogelito just an idea: I think a way to avoid race conditions using transactions is with row-versioning. This need that [Update Document](https://appwrite.io/docs/server/databases?sdk=nodejs-default#databasesUpdateDocument) has a "queries" parameter (or at least a version param).

For instance: A document has a timestamp field/column called version (it changes each time document is updated), this field can optionally be created by the developer. Inside a transaction, the document is firstly Read (with version X1), then some fields are increased or changed, then updateDocument is called, specifying the queries args (Query.equal("version", X1)) so if the record was not found by where clause any update will be done and an exception should be thrown. That means that the document was modified between the read and the update. The developer should choose a return the error or retry.

**Comment by fogelito** [comment]
@nev-21 Hmmm,  using "versioning,"  I think will be more complicated, and will require an extra field and query for each insert, it will be a cumbersome operation,  we would rather use the built-in mechanism of the database, which is transactions (Isolation levels).

**Comment by nev-21** [comment]
@fogelito Yeah I'm agree it is easier to use Isolation levels, it would be great if developers can specify the isolation level when the transaction starts (to ensure flexibility).

**Comment by therealsras** [comment]
is there any update about this issue

**Comment by barart** [comment]
+1 on this, i hope it comes onboard with 1.5

**Comment by burkle3** [comment]
+1 on this.

**Comment by idan054** [comment]
+1

**Comment by haysquareA** [comment]
And here we are 2024, and the feature is still unavailable

**Comment by morgueye4** [comment]
This feature would be a game changer

**Comment by dennislbc** [comment]
Still no upates on this?

**Comment by LDY681** [comment]
They add this feature to the 1.7.0 roadmap, hope it won't get postponed lol

**Comment by soranoo** [comment]
+1

**Comment by Khalidhassan3011** [comment]
still waiting

**Comment by DefacedFace** [comment]
me, in 2025, too lazy to get doc value and update it

**Comment by shubshinde** [comment]
+1

**Comment by FWHD** [comment]
I don't mind if it's .increment or add or something, but this would be really helpful. Would be great not to have to fetch the data either as part of the logic to update the document or pre-emptively (and then having to propagate it in parts of the app where it's not needed).

**Comment by RoosterInc** [comment]
It's 2025 and there is no update on this 🫢

**Comment by minecrawler** [comment]
@RoosterInc wdym? It was closed three weeks ago. Just look at the GitHub update.

<img width="468" height="249" alt="Image" src="https://github.com/user-attachments/assets/1f2a9eb2-aafd-476c-a8b2-3cc174996e30" />

It'll be part of Appwrite v1.8 (Cloud is currently on 1.7). 1.7 was released in May, and from previous releases we can estimate that new minor releases happen every half a year (maybe a little longer), so 1.8 will probably ship with this feature in 2025Q4 to 2026Q1.
